handler.js

```diff
@@ -0,0 +1,103 @@
+
+  const sns = new AWS.SNS();
+  const snsParams = {
+    Message: JSON.stringify(result),
+    Subject: "SNS from Transaction Lambda",
+    TopicArn: topicArn
+  };
+  sns.publish(snsParams, context.done);
+}
+
+module.exports.transaction = async (event, context) => {
+  const actionSNS = event.Records[0].Sns.TopicArn;
+
+  // SNS has single record only https://stackoverflow.com/questions/33690231/when-lambda-is-invoked-by-sns-will-there-always-be-just-1-record
+  // requires jobId in Message
+  const jobId = event.Records[0].Sns.Message;
+  const serviceBase = process.env.KEY_SERVICE_BASE_URL || 'http://pair.service.thinblock.io/';
+  const jobBase = process.env.JOBS_SERVICE_BASE_URL || 'http://jobs.service.thinblock.io/';
+  const result = {
+    jobId,
+    event: '',
+    data: ''
+  }
+
+  const jobURL = `${jobBase}jobs/${jobId}?look_up=_id`;
+  const { data: { actions } } = await axios.get(jobURL);
+
+  // find appropriate action in job
+  let actionIndex = -1;
+  actions.forEach((element, index) => {
+    if (element.action.sns_topic_arn === actionSNS) {
+      actionIndex = index;
+    }
+  });
+
+  if (actionIndex = -1) {
+    // Couldn't find appropriate action in job
+    result.event = null;
+    result.data = 'transaction_failed: Action not found in job';
+    publishSNS(result);
+    return;
+  }
+
+  const { params, action } = actions[actionIndex];
+  result.event = action.name;
+
+  if (params.user_id === undefined || params.exchange === undefined ||
+    params.symbol === undefined || params.side === undefined ||
+    params.amount === undefined || params.price === undefined) {
+    result.data = 'transaction_failed: Requires user_id, exchange, symbol, side, amount, price'
+    publishSNS(result);
+    return;
+  }
+
+  const user = params.user_id;
+  const exchangeName = params.exchange;
+  let { symbol, side, amount, price } = params;
+  let order = null;
+
+  try {
+    const serviceURL = `${serviceBase}pairs?user_id=${user}&exchange=${exchangeName}`;
+    const { key, secret } = (await axios.get(serviceURL)).data;
+
+    const exchangeId = exchangeName
+    , exchangeClass = ccxt[exchangeId]
+    , exchange = new exchangeClass ({
+      'apiKey': key,
+      'secret': secret,
+      'timeout': 30000,
+      'enableRateLimit': true,
+    });
+
+    await exchange.loadMarkets();
+    amount = exchange.amountToPrecision(symbol, amount);
+    price = exchange.priceToPrecision(symbol, price);
+
+    if (side === 'buy') {
+      order = await exchange.createLimitBuyOrder(symbol, amount, price, { test: true });
+    } else if (side === 'sell') {
+      order = await exchange.createLimitSellOrder(symbol, amount, price, { test: true });
+    } else {
+      result.data = 'transaction_failed: Side should be "buy" or "sell"';
+      publishSNS(result);
+      return;
+    }
+
+    result.data = 'transaction_completed: Success';
```

Comment:
```javascript
result.data = {
  status: 'success',
  message: 'transaction_completed: success',
  data: {
    tx_id: 'tx id from binance?'
  }
}
```

Reply:
We don't know what will be returned from the exchange at this point, since I've only tested it with `test` flags and an empty account. There will be a few other responses from the exchange which we'll find out after actual testing (e.g. low balance, not allowed, etc.). We can categorize those later then?

Reply:
Okay sure, i think we can log the `order` then so we can find out?

Reply:
👍